### ISSUE TITLE
Best practices with Multi-GPU VideoReader and sharding

### ISSUE THREAD

**Issue by Timen** [NVIDIA/DALI#2310]
After having solved the #2296 of using numpy arrays as labels when using the VideoReader for loading data, I had some questions regarding the use and best practices which the documentation left a little ambiguous. The questions are as followed:

1. From what I have seen it is not yet possible to use the VideoReader with serialized data like Tfrecords, only with videos stored as mp4 files on the drive correct?
2. What video format is the most efficient to decode, I'm currently using libx264 with yuv420 pixel format (as this was recommended in the old NVVL repo)
3. How does a sharded videoreader work when integrated in a torch.utils.data.Dataset? To be more precise I have this code in the init function of a torch.utils.data.Dataset object:

```
self.pipe = VideoPipe(batch_size=batch_size, num_threads=2, device_id=shard_id, data=path_to_file,sequence_length=num_frames,num_shards=num_shards,shuffle=shuffle,drop_last=drop_last)
self.pipe.build()
pipes = [self.pipe]
self.dali_iter = DALIClassificationIterator(pipes, pipes[0].epoch_size("Reader"))
```

Where :
```
shard_id =  torch.distributed.get_rank()
num_shards = torch.distributed.get_world_size()
```

I can see this dataloader getting initialized 4 times (using 4 GPUS) with each a unique shard_id (0 to 3) and the same number of shards (namely 4). The iterator is further implemented as followed:
```
def __len__(self):
        return int(self.epoch_size)//self.batch_size//self.num_shards
def __iter__(self):
        return self
def __next__(self):
        data = self.dali_iter.__next__()
        #some code to generate labels
        return frames,labels
```
And while (to my surprise) this works fine, there is an issue with the number of iterations reported for an epoch, vs the actual number of iterations the training code runs. When sharding the data in the VideoReader my assumption was that this divides the dataset in 4 unique shards each containing a quarter of the total examples in that dataset (as is reflected in the __len__ function). However when using this setup to train, the number of iterations the training runs for is actually 4 times the number of iterations it should be running for, to process 1 epoch. I'm unsure how this is possible as the length is correctly reported. This may be more of a pytorch question vs a DALI specific one, but as the documentation on multi-gpu video reader is not available I figure I also ask it here, to make sure I didn't miss understood any of the available documentation.

**Comment by JanuszL** [comment]
Hi,
> From what I have seen it is not yet possible to use the VideoReader with serialized data like Tfrecords, only with videos stored as mp4 files on the drive correct?

That is correct. The VideoReader was designed to work with the Files not in the memory buffer. There is nothing that would prevent such functionality with the current design. We will add it to our backlog but we cannot commit to any particular date.

> What video format is the most efficient to decode, I'm currently using libx264 with yuv420 pixel format (as this was recommended in the old NVVL repo)

I would check the performance comparison of the VideoSDK [itself here](https://developer.nvidia.com/nvidia-video-codec-sdk).

> How does a sharded videoreader work when integrated in a torch.utils.data.Dataset? To be more precise I have this code in the init function of a torch.utils.data.Dataset object:

I would recommend using `reader_name` argument available in the nightly build that would the Iterator infer the necessary parameters directly from the reader. If you want to use the size argument please, as the documentation states `Number of samples in the shard for the wrapped pipeline (if there is more than one it is a sum)`, provide the `int(self.epoch_size)//self.num_shards` value.

**Comment by Timen** [comment]
@JanuszL thank you for your quick and clear reply, I'll check out the nightly build and re-encode the videos as HVEC 4:2:0 as that seems to be the right one for my setup. I'm not running into an I/O bottleneck currently with reading the mp4 files from SSD but as the datasets are pretty large an HDD friendly serialized data format would be great in the future!
